src/state.rs

```diff
@@ -122,40 +126,103 @@ impl StateContainer {
-                need_init = true;
-            }
+            self.prepare_mmap(create)?;
+        }
 
-            let mut mmap = unsafe { MmapMut::map_mut(&state_file).map_err(FailedStateRead)? };
-            if need_init {
-                // Create new state and write it to mapped memory
-                let new_state = State::default();
-                unsafe {
-                    mmap.as_mut_ptr()
-                        .copy_from((&new_state as *const State) as *const u8, STATE_SIZE)
-                };
-            };
+        let mmap = self.mmap.as_ref().unwrap().read();
 
-            self.mmap = Some(mmap);
+        Ok(unsafe { &mut *(mmap.as_ptr() as *mut State) })
+    }
+
+    /// Fetch state from existing memory mapped file or create new one.
+    ///
+    /// If write locking is enabled, acquire the lock before initializing memory, returning a
+    /// lock conflict error if the lock cannot be acquired.
+    #[inline]
+    pub(crate) fn state_write(
+        &'a mut self,
+        create: bool,
+    ) -> Result<&'a mut State, SynchronizerError> {
+        if self.mmap.is_none() {
+            self.prepare_mmap(create)?;
         }
-        Ok(unsafe { &mut *(self.mmap.as_ref().unwrap().as_ptr() as *mut State) })
+
+        let mut guard = self.mmap.as_mut().unwrap().lock()?;
+        let mmap = guard.deref_mut();
+
+        Ok(unsafe { &mut *(mmap.as_ptr() as *mut State) })
+    }
+
+    /// Initialize mmaped memory from the state file.
+    #[inline]
+    pub(crate) fn prepare_mmap(&mut self, create: bool) -> Result<(), SynchronizerError> {
+        let mut opts = OpenOptions::new();
+        opts.read(true).write(true).create(create);
+
+        // Only add mode on Unix-based systems to allow read/write from owner/group only
+        #[cfg(unix)]
+        opts.mode(0o660);
+
+        let state_file = opts.open(&self.state_path).map_err(FailedStateRead)?;
+
+        let mut need_init = false;
+        // Reset state file size to match exactly `STATE_SIZE`
+        if state_file.metadata().map_err(FailedStateRead)?.len() as usize != STATE_SIZE {
+            state_file
+                .set_len(STATE_SIZE as u64)
+                .map_err(FailedStateRead)?;
+            need_init = true;
+        }
+
+        let mut mmap = unsafe { MmapMut::map_mut(&state_file).map_err(FailedStateRead)? };
+        if need_init {
+            // Create new state and write it to mapped memory
+            let new_state = State::default();
+            unsafe {
+                mmap.as_mut_ptr()
+                    .copy_from((&new_state as *const State) as *const u8, STATE_SIZE)
+            };
+        };
+
+        self.mmap = Some(WL::new(mmap, state_file));
+        Ok(())
+    }
+}
+
+#[cfg(all(test, unix))]
+mod tests {
+    use super::*;
+    use crate::locks::SingleWriter;
+    use crate::synchronizer::SynchronizerError;
+
+    #[test]
+    fn single_writer_lock_mode_prevents_duplicate_writer() {
```

Comment:
Could we also test with two different synchronizers?

Reply:
Yeah, I meant two synchronizers with same lock strategy.

Reply:
Added a test for this in `synchronizer.rs`.